### PR TITLE
Improve documentation and implementation of seismic anomalies post-processor

### PR DESCRIPTION
**Pull request by bobmyhill**
This PR addresses the problems with the seismic anomalies post-processor outlined in Issue #96 by:
* Adding n_slices as a user-defined parameter
* Improving the description of the post-processor, and increasing the number of comments
* Changing the processing of the lateral average calculation. The old code evaluated the average Vp and Vs in 50 depth slices, then took a 3-point simple running average of those values, and used that average for all cells within each depth range. The new code now evaluates the average Vp and Vs in n depth slices, then interpolates linearly between these values at the depths of each cell. This avoids systematic under- and overshooting at the edges of each depth slice.

### CODE REVIEW

File: include/aspect/postprocess/visualization/seismic_anomalies.h [hunk 1]
```diff
@@ -61,6 +61,23 @@ namespace aspect
           virtual
           std::pair<std::string, Vector<float> *>
           execute () const;
+
+          /**
+                 * Declare the parameters this class takes through input files.
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/seismic_anomalies.h]: indent these lines please (unfortunately astyle does not fix comment lines).

File: include/aspect/postprocess/visualization/seismic_anomalies.h [hunk 2]
```diff
@@ -61,6 +61,23 @@ namespace aspect
...
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int n_slices;
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/seismic_anomalies.h]: Document this parameter please.

File: include/aspect/postprocess/visualization/seismic_anomalies.h [hunk 3]
```diff
@@ -93,6 +110,23 @@ namespace aspect
           virtual
           std::pair<std::string, Vector<float> *>
           execute () const;
+
+          /**
+                 * Declare the parameters this class takes through input files.
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/seismic_anomalies.h]: Indentation

File: include/aspect/postprocess/visualization/seismic_anomalies.h [hunk 4]
```diff
@@ -93,6 +110,23 @@ namespace aspect
...
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int n_slices;
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/seismic_anomalies.h]: Documentation please :smile:

File: source/postprocess/visualization/seismic_anomalies.cc [hunk 5]
```diff
@@ -68,17 +40,24 @@ namespace aspect
         return_value ("Vs_anomaly",
                       new Vector<float>(this->get_triangulation().n_active_cells()));
 
+        // Calculate the Vs averages within n slices
+        std::vector<double> Vs_depth_average(n_slices);
+        this->get_lateral_averaging().get_Vs_averages(Vs_depth_average);
+
+        // Pad averages
```
Comment (gassmoeller) [inline source/postprocess/visualization/seismic_anomalies.cc]: Can you quickly explaine in the comment why you pad?

File: source/postprocess/visualization/seismic_anomalies.cc [hunk 6]
```diff
@@ -116,22 +99,32 @@ namespace aspect
...
+              // the depth slice which has its center just above that point
               const double depth = this->get_geometry_model().depth(fe_values.quadrature_point(0));
-              const unsigned int idx = static_cast<unsigned int>((depth*num_slices)/max_depth);
-              Assert(idx<num_slices, ExcInternalError());
 
-              // compute the deviation from the average in per cent
-              (*return_value.second)(cell_index) = (Vs - Vs_depth_average[idx])/Vs_depth_average[idx]*1e2;
+              // N.B. static_cast<int> always truncates the double (i.e. 1.9 -> 1)
+              // Here we subtract 0.5 to find the first slice with its center shallower than the cell,
+              // and add 1 to take padding into account
```
Comment (gassmoeller) [inline source/postprocess/visualization/seismic_anomalies.cc]: I do not see an add 1 in the code?

File: source/postprocess/visualization/seismic_anomalies.cc [hunk 7]
```diff
@@ -116,22 +99,32 @@ namespace aspect
...
+
+              // Find the depth of the zeroth quadrature point in the cell and work out
+              // the depth slice which has its center just above that point
               const double depth = this->get_geometry_model().depth(fe_values.quadrature_point(0));
-              const unsigned int idx = static_cast<unsigned int>((depth*num_slices)/max_depth);
-              Assert(idx<num_slices, ExcInternalError());
 
-              // compute the deviation from the average in per cent
-              (*return_value.second)(cell_index) = (Vs - Vs_depth_average[idx])/Vs_depth_average[idx]*1e2;
+              // N.B. static_cast<int> always truncates the double (i.e. 1.9 -> 1)
```
Comment (gassmoeller) [inline source/postprocess/visualization/seismic_anomalies.cc]: What does N.B. stand for?

File: source/postprocess/visualization/seismic_anomalies.cc [hunk 8]
```diff
@@ -140,17 +133,24 @@ namespace aspect
         return_value ("Vp_anomaly",
                       new Vector<float>(this->get_triangulation().n_active_cells()));
 
+        // Calculate the Vp averages within n slices
+        std::vector<double> Vp_depth_average(n_slices);
+        this->get_lateral_averaging().get_Vp_averages(Vp_depth_average);
+
+        // Pad averages
```
Comment (gassmoeller) [inline source/postprocess/visualization/seismic_anomalies.cc]: same as above

File: source/postprocess/visualization/seismic_anomalies.cc [hunk 9]
```diff
@@ -188,25 +192,121 @@ namespace aspect
...
+
+              // Find the depth of the zeroth quadrature point in the cell and work out
+              // the depth slice which has its center just above that point
               const double depth = this->get_geometry_model().depth(fe_values.quadrature_point(0));
-              const unsigned int idx = static_cast<unsigned int>((depth*num_slices)/max_depth);
-              Assert(idx<num_slices, ExcInternalError());
 
-              // compute the deviation from the average in per cent
-              (*return_value.second)(cell_index) = (Vp - Vp_depth_average[idx])/Vp_depth_average[idx]*1e2;
+              // N.B. static_cast<int> always truncates the double (i.e. 1.9 -> 1)
```
Comment (gassmoeller) [inline source/postprocess/visualization/seismic_anomalies.cc]: same as above

**Comment by bobmyhill** [comment]
Thanks @gassmoeller , I've addressed all your points. Hopefully this should work better now.

**Comment by bobmyhill** [comment]
Well, this grew a little bit! I added a couple more comments regarding choice of values for the model parameters, ran astyle (oops, must remember this next time) and checked and updated the output. 

For the steinberger_seismic_velocities example, we get about a factor of three improvement in lateral velocity estimation. I suspect that this is largely down to removing the running average. In larger models, the performance should be even better.

File: include/aspect/postprocess/visualization/seismic_anomalies.h [hunk 10]
```diff
@@ -93,6 +115,28 @@ namespace aspect
           virtual
           std::pair<std::string, Vector<float> *>
           execute () const;
+
+          /**
+          * Declare the parameters this class takes through input files.
+          */
```
Comment (bangerth) [inline include/aspect/postprocess/visualization/seismic_anomalies.h]: same here

File: include/aspect/postprocess/visualization/seismic_anomalies.h [hunk 11]
```diff
@@ -93,6 +115,28 @@ namespace aspect
...
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          /**
+           * Number of depth slices used to define average
+           * seismic compressional wave velocities from which anomalies
+           * are calculated.
+           **/
```
Comment (bangerth) [inline include/aspect/postprocess/visualization/seismic_anomalies.h]: and here

File: tests/steinberger_seismic_velocities.prm [hunk 12]
```diff
@@ -86,12 +86,18 @@ subsection Postprocess
...
+    subsection Vs anomaly
+      set Number of depth slices = 96
+    end
   end
 
   subsection Depth average
     set Time between graphical output = 0
     set Number of zones = 10
- end
+  end
```
Comment (bangerth) [inline tests/steinberger_seismic_velocities.prm]: nice catch :-)

**Comment by gassmoeller** [comment]
This PR now has conflicts with master, sorry for that. Ask @jdannberg if you need help resolving them, I think she is responsible for the problem :smile:. Otherwise ready to merge when the conflicts are fixed.

**Comment by gassmoeller** [comment]
Please leave me a note when this is ready from your side.

**Comment by bobmyhill** [comment]
This is ready from my side! :)

**Pull request merged**